perf-tool/src/exception.cpp

```diff
@@ -72,53 +71,62 @@ JNIEXPORT void JNICALL Exception(jvmtiEnv *jvmtiEnv,
     jvmtiLineNumberEntry *lineTable;
     char buffer[20];
     char *methodName;
+    int numExceptions;
 
-    static int numExceptions = 0; // number of total exceptions recorded
-    numExceptions++;
+    /* Get number of exceptions recorded and increment */
+    numExceptions = atomic_fetch_add(&exceptionSampleCount, 1);
+    jdata["numExceptions"] = numExceptions;
 
-    // First get exception address
-    char ptrStr[] = "%p";
-    sprintf(buffer, ptrStr, exception);
-    string exceptionStr(buffer);
-    jdata["exceptionAddress"] = (char*)exceptionStr.c_str();
+    /* Get exception address */
+    sprintf(buffer, "%p", exception);
+    jdata["exceptionAddress"] = buffer;
 
     /* Get method name */
     err = jvmtiEnv->GetMethodName(method, &methodName, NULL, NULL);
-    jdata["callingMethod"] = (char *)methodName;  // record calling method
+    if (err == JVMTI_ERROR_NONE) {
+        jdata["callingMethod"] = methodName;  // record calling method
+    }
+    err = jvmtiEnv->Deallocate((unsigned char*)methodName);
+    check_jvmti_error(jvmtiEnv, err, "Unable to deallocate methodName");
```

Comment:
If `getMethodName` failed, then probably there was no method allocated, so when we try to deallocate, we'll have another failure and bring down the JVM in `check_jvmti_error`